### DISCUSSION TITLE
Error: Failed to load resource: the server responded with a status of 413 ()

### DISCUSSION THREAD

**Discussion by ScottAgirs** [vercel/next.js#64647] [Ideas]
### What version of Next.js are you using?

^10.0.9

### What version of Node.js are you using?

14.0.0

### What browser are you using?

Chrome, Brave, Safari

### What operating system are you using?

macOS Big Sur

### How are you deploying your application?

Vercel

### Describe the Bug

Passing a file from Next.js client to Next.js serverless function in `pages/api` keeps failing w status 413.

Locally, in development all works as expected, but when deployed throws following error:
<img width="756" alt="Screen Shot 2021-05-29 at 9 48 55 PM" src="https://user-images.githubusercontent.com/11737572/120089386-ae495b00-c0c7-11eb-890a-21bf75c83ee9.png">
`Failed to load resource: the server responded with a status of 413 ()`

I tried to strip everything down to debug, and ended up with only this in my serverless function, that yields the same result:
```javascript
// /pages/api/video/upload.js
export const config = {
  api: {
    bodyParser: false,
  },
};

export default async (req, res) => {
  console.log('req.body', req.body);
};

```

I then call it from the client like so:
```javascript
const formData = new FormData();
formData.append('file', file);

const response = await fetch(`/api/video/upload`, {
        body: formData,
        method: 'POST',
      }).catch(err => console.log('err 0', err));
```

### Expected Behavior

Be able to pass formData with a file in deployed to production on Vercel same way as it works in development, locally.

### To Reproduce

[Here is a Codesandbox](https://codesandbox.io/s/floral-frost-jnkwc?file=/pages/index.js), which has the same issue - throws 413.

Here's the punchline:
### **The same exact code works on developmenttttt.**
<sub>why, Next.js/vercel gawds, whyy</caption>

**Comment by timneutkens** [comment]
It seems that you are sending more than 5MB in API route, which exceeds this limit: https://vercel.com/docs/platform/limits#serverless-function-payload-size-limit
We could potentially limit this automatically for API routes created in Next.js.

**Comment by ScottAgirs** [comment]
@timneutkens thanks for pointing that out!

What puzzles me is that seemingly [exact same setup](https://codesandbox.io/s/brave-moon-9o66b?from-embed=&file=/pages/api/upload.js) by @tpiros works just fine, and it also works on my local env.

Am I missing something? 🙈

**Comment by tpiros** [comment]
@ScottAgirs I did note the same behaviour in my setup (see: https://tpiros.dev/blog/uploading-and-displaying-videos-with-nextjs/) but I thought it was a limitation on Codesandbox's side. 

@timneutkens maybe also take the filetype into consideration? As per the example, video uploads may become a common operation as well and 5MB for a video is an impossible limit to match. For videos (and for other large binaries, should you choose to accept those) you could do chunked uploads.

**Comment by ScottAgirs** [comment]
Hi @timneutkens is what [Tamas mentioned above](https://github.com/vercel/next.js/issues/25598#issuecomment-851337855) something that could be considered and if so any estimates on when could expect this?

I also believe that it could become a common operation, additionally, this is a core need for my use-case and presumably for many others too.

🤞🏼

**Comment by timneutkens** [comment]
5MB is an upstream limit practically all serverless function providers have, so it's unlikely it'll be lifted anytime soon. For processing videos you'll want to use an external service like https://mux.com/.

**Comment by tpiros** [comment]
Cloudinary has video processing capabilities as well but I guess the question is, how to get these videos to any of these service providers at the first place if the videos are larger than 5MB.

**Comment by CrocoDillon** [comment]
@timneutkens sorry for bumping this issue but since the cause and reasoning is clear I really didn't want to open a new one for my question.

We recently moved our GraphQL backend to our Next.js stack to streamline development (it's just me and my brother working on this project in our spare time so anything to simplify the process helps a lot).

However part of the GraphQL API is uploading images directly from a (mobile) device's camera. These images are more often than not over 5MB. I use graphql-upload for getting images out of the GraphQL request, then determine image dimensions and store the image in Google Cloud Storage. Storing directly to Google Cloud Storage complicates the setup and won't give me the image dimensions.

Any tips to handle this? I'd hate to have to take the GraphQL backend out of the Next.js stack again. Is the limit only for API routes? Can I maybe POST to a normal route?

Edit: As a side note, before moving the GraphQL backend we did have a proxy API route (to handle authorization headers) and didn't have any issues uploading images.

**Comment by wyattjoh** [comment]
The best path I've seen for this class of uploads is to send them directly to object storage solutions (like Google Cloud Storage, Amazon S3, Backblaze B2) via a [signed URL](https://docs.aws.amazon.com/AmazonS3/latest/userguide/PresignedUrlUploadObject.html). You could generate the signed URL inside an API route, and upload the file directly to the object storage. After the file was uploaded you can call back either via an event on the cloud storage side to your Next.js app or have your Next.js client side call back when the file upload is complete.

**Comment by CrocoDillon** [comment]
> The best path I've seen for this class of uploads is to send them directly to object storage solutions (like Google Cloud Storage, Amazon S3, Backblaze B2) via a [signed URL](https://docs.aws.amazon.com/AmazonS3/latest/userguide/PresignedUrlUploadObject.html). You could generate the signed URL inside an API route, and upload the file directly to the object storage. After the file was uploaded you can call back either via an event on the cloud storage side to your Next.js app or have your Next.js client side call back when the file upload is complete.

That's only an option if you don't need any processing (like extracting dimensions for the Image component) isn't it?

**Comment by wyattjoh** [comment]
While there is a limit of how large a payload you can send to an edge function, there is no such limit as to how much an edge function can request for itself! You should be able to instruct the edge function to download the file from cloud storage to get process it. Alternatively, you can have another lambda or other function perform that operation within the cloud and send the results back to the edge function.
